### PR TITLE
rfcs: parser as a go sub-module

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Problem Summary: The formal RFC for the draft https://internals.tidb.io/t/topic/385.

### Release note 

```release-note
None
```

### CODE REVIEW

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (kennytm) [inline docs/design/2021-09-13-parser-as-submodule-of-tidb.md:59]: (issues can be transferred, no need to close and reopen)

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (kennytm) [inline docs/design/2021-09-13-parser-as-submodule-of-tidb.md:83]: or we could use true semver for the parser sub-module instead of tying with the product release cycle

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (kennytm) [inline docs/design/2021-09-13-parser-as-submodule-of-tidb.md:87]:
```suggestion
We can not simply remove it. Importing pingcap/parser to build applications needs parser.go file, but go module lacks build scripts in rust or `postbuild` in nodejs to auto-generate parser.go.
```

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (xhebox) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md:83]: I would like to leave that as another RFC.

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (xhebox) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md:59]: https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository, this, right? I know...

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (xhebox) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md:83]: And it is probably a good idea to keep identical version with tidb, before it becomes a real "standalone" parser.

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (xhebox) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md]: Updated.

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (bb7133) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md:83]: Yes, IMHO semver is a bigger topic than this, considering all components of TiDB & other components.

**Comment by tangenta** [comment]
Here is a rough comparison between the master branch of TiDB and parser, FYI:

|    | parser | TiDB |
| --- | --- | --- |
| cloc (line of code) | 68124  | 869372 |
| commits | 750 | 14571 |
| size of git clone | 59MB | 181MB |

I don't think git bloating will be a problem.

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (tangenta) [inline docs/design/2021-09-13-parser-as-submodule-of-tidb.md:1]:
```suggestion
# Proposal: Move parser back to pingcap/tidb
```

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (tisonkun) [inline docs/design/2021-09-13-parser-as-submodule-of-tidb.md:58]:
```suggestion
2. Stop sending new PRs and issues to pingcap/parser, new PRs should be directly sent to pingcap/tidb. And merge recent PRs in pingcap/parser as much as possible. Changes will be synchronized from pingcap/parser to pingcap/tidb manually, of course.
```

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (tisonkun) [inline docs/design/2021-09-13-parser-as-submodule-of-tidb.md:62]: You should also mention how to migrate the test infra.

Comment (tisonkun) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md:62]: OK I see it below, but please consider it in the implementation plan. It is ended by migrated, not only answered the question.

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (tisonkun) [inline docs/design/2021-09-13-parser-as-submodule-of-tidb.md:87]:
```suggestion
We can not simply remove it. Importing pingcap/parser to build applications needs parser.go file, but go module doesn't support something like build scripts in rust or `postbuild` in nodejs to auto-generate parser.go.
```

**Review by tisonkun — COMMENTED** [review]
Thanks for creating the RFC. Comments inline.

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (xhebox) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md:60]: Added in step 3.

**Review by wjhuang2016 — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

File: docs/design/2021-09-13-parser-as-submodule-of-tidb.md
Comment (tisonkun) [inline docs/design/2021-09-13-parser-as-submodule-of-tidb.md:5]: @xhebox please create an tracking issue before merging this RFC.

Comment (xhebox) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md:5]: I won't merge it now, since I've promised to have 3 LGTM and wait for 3 days. I would create one later, you could give the 3rd LGTM.

Comment (tisonkun) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md:5]: OK. I'll review this proposal today.

Comment (xhebox) [reply docs/design/2021-09-13-parser-as-submodule-of-tidb.md]: #28257, here it is.

**Review by tisonkun — APPROVED** [review]
LGTM for the motivation and detailed plan. Release and versioning could be as is for now.

**Review by bb7133 — APPROVED** [review]

**Review by zhouqiang-cl — APPROVED** [review]
LGTM

**Review by zhouqiang-cl — APPROVED** [review]
LGTM

**Comment by xhebox** [comment]
It is been 3 days without objection, merge it.
/merge

**Pull request merged**
